### ISSUE TITLE
Any idea on how to use nestjs cli plugin with nx?

### ISSUE THREAD

**Issue by JeetChaudhari** [nrwl/nx#4135]
I want to use nestjs cli plugin as mentioned at https://docs.nestjs.com/graphql/cli-plugin#using-the-cli-plugin. As my nx workspace does not contain nest-cli.json, I am not sure how to use this.

Also If I manually create nest-cli.json file it seems to have no effect. At https://nx.dev/latest/node/workspace/configuration#schematics it is mentioned that we can configure options from here but I am not sure how to do it.

Can anyone help me with this please? One more thing, I could not add question here as slack link is broken on add new issue page.

**Comment by JeetChaudhari** [comment]
I was able to solve this using https://github.com/nrwl/nx/issues/2147#issuecomment-707842322, Just replaced swagger with graphql as 
from `return { before: [require('@nestjs/swagger/plugin').before({}, program)] };`
to `return { before: [require('@nestjs/graphql/plugin').before({}, program)] };`
saved file as "webpack.config.js" and added it to "workspace.json" as 

/workspace.json
``` 
"projects": {
  "api": {
    "architect": {
      "build": {
        "options": {
          "webpackConfig": "apps/api/webpack.config.js"
        }
      }
    }
  }
}
```

/apps/api/webpack.config.js
```
module.exports = (config, _context) => {
  const tsLoader = config.module.rules.find((r) =>
    r.loader.includes("ts-loader")
  );

  if (tsLoader) {
    tsLoader.options.transpileOnly = false;
    tsLoader.options.getCustomTransformers = (program) => {
      return {
        before: [
          require("@nestjs/graphql/plugin").before(
            {
              // Options: typeFileNameSuffix, introspectComments
            },
            program
          ),
        ],
      };
    };
  }

  return config;
};

```

**Comment by crohit92** [comment]
I created my workspace with nx and was looking for the same issue.
I got it solved by adding 
```
"tsPlugins": [
    {
      "name": "@nestjs/swagger/plugin",
      "options": {
        "classValidatorShim": false,
        "introspectComments": true
      }
    }
 ]
```

in my angular.json (Since my workspace was NestJS + Angular)
under projects->api->architect->build->options

**Comment by kirillvarikow** [comment]
> I created my workspace with nx and was looking for the same issue. I got it solved by adding
> 
> ```
> "tsPlugins": [
>     {
>       "name": "@nestjs/swagger/plugin",
>       "options": {
>         "classValidatorShim": false,
>         "introspectComments": true
>       }
>     }
>  ]
> ```
> 
> in my angular.json (Since my workspace was NestJS + Angular) under projects->api->architect->build->options

Hi! 
I don't understand where to put this piece of code? 
Could you explain it, please? 
Should I put it into my `nx.json` ? Or it should be something else ? Thanks

**Comment by aviggiano** [comment]
hi @kirillvarikow and others with this issue

You should put it under the nx `project.json` file

Like this, entire file:

```typescript
{
  "root": "packages/backend",
  "sourceRoot": "packages/backend/src",
  "projectType": "application",
  "targets": {
    "build": {
      "executor": "@nrwl/node:webpack",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/backend",
        "main": "packages/backend/src/main.ts",
        "tsConfig": "packages/backend/tsconfig.app.json",
        "tsPlugins": [
          {
            "name": "@nestjs/swagger/plugin",
            "options": {
              "classValidatorShim": false,
              "introspectComments": true
            }
          }
        ],
        "assets": ["packages/backend/src/common/assets"]
      },
      "configurations": {
        "production": {
          "optimization": true,
          "extractLicenses": true,
          "inspect": false
        }
      }
    },
    "serve": {
      "executor": "@nrwl/node:node",
      "options": {
        "buildTarget": "backend:build"
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/backend/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/backend"],
      "options": {
        "jestConfig": "packages/backend/jest.config.js",
        "passWithNoTests": true,
        "codeCoverage": true,
        "coverageDirectory": "coverage/packages/backend"
      }
    }
  },
  "tags": []
}

```

**Comment by KevinOsorioCodes** [comment]
HI! the @nrwl/nest plugin has the nest cli built in, you just have to use the commands like this:
`nx g @nrwl/nest:<commandName> <finalName> --project=<projectName>`

ex:  nx g @nrwl/nest:resource client --project=api


Here the docs: https://nx.dev/packages/nest#nest-generators


I hope it works for you!

**Comment by eliotis** [comment]
> I was able to solve this using [#2147 (comment)](https://github.com/nrwl/nx/issues/2147#issuecomment-707842322), Just replaced swagger with graphql as from `return { before: [require('@nestjs/swagger/plugin').before({}, program)] };` to `return { before: [require('@nestjs/graphql/plugin').before({}, program)] };` saved file as "webpack.config.js" and added it to "workspace.json" as
> 
> /workspace.json
> 
> ```
> "projects": {
>   "api": {
>     "architect": {
>       "build": {
>         "options": {
>           "webpackConfig": "apps/api/webpack.config.js"
>         }
>       }
>     }
>   }
> }
> ```
> 
> /apps/api/webpack.config.js
> 
> ```
> module.exports = (config, _context) => {
>   const tsLoader = config.module.rules.find((r) =>
>     r.loader.includes("ts-loader")
>   );
> 
>   if (tsLoader) {
>     tsLoader.options.transpileOnly = false;
>     tsLoader.options.getCustomTransformers = (program) => {
>       return {
>         before: [
>           require("@nestjs/graphql/plugin").before(
>             {
>               // Options: typeFileNameSuffix, introspectComments
>             },
>             program
>           ),
>         ],
>       };
>     };
>   }
> 
>   return config;
> };
> ```

Hey JeetChaudhari, is this solution still working for you because for me after nx ver 14 it stopped. Any help?

**Comment by eliotis** [comment]
Is this solution for @nestjs/graphql/plugin  still working for you because for me after nx ver 14 it stopped. 

Any help?

**Comment by tagplus5** [comment]
Example project.json with @nestjs/swagger plugin

```
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "apps/api/src",
  "projectType": "application",
  "targets": {
    "build": {
      "executor": "@nrwl/node:webpack",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/apps/api",
        "main": "apps/api/src/main.ts",
        "tsConfig": "apps/api/tsconfig.app.json",
        "tsPlugins": [
          "@nestjs/swagger/plugin"
        ],
        "assets": [
          "apps/api/src/assets"
        ],
        "generatePackageJson": true
      },
      "configurations": {
        "production": {
          "optimization": true,
          "extractLicenses": true,
          "inspect": false,
          "fileReplacements": [
            {
              "replace": "apps/api/src/environments/environment.ts",
              "with": "apps/api/src/environments/environment.prod.ts"
            }
          ]
        }
      }
    },
    "serve": {
      "executor": "@nrwl/node:node",
      "options": {
        "buildTarget": "api:build"
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["apps/api/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/apps/api"],
      "options": {
        "jestConfig": "apps/api/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
```

**Comment by dan-cooke** [comment]
TS plugins has been removed in the latest version of NX without any notification, so this is now broken.

#12966 

The only option is to use a custom webpack config, which is messy

**Comment by mattvgm** [comment]
@dan-cooke Replace `tsPlugins` to `transformers` now. It looks like it works the same way, just different name

**Comment by dan-cooke** [comment]
Yeah I found that out via opening another issue, thanks!

**Comment by eso32** [comment]
TLDR

your configuration in project.json should look like this:
`
{
    "name": "project-with-swagger-plugin",
    "targets": {
        "build": {
            "executor": "@nrwl/webpack:webpack",
            "options": {
                // other options 
                "transformers": [
                    // register plugins here
                    "@nestjs/swagger/plugin"
                ]
            },
        }
   }
}
`

Thanks @mattvgm

**Comment by MartinDevillers** [comment]
Thank you @eso32 
One addition for those looking to change any of the plugins options:
```
{
    "name": "project-with-swagger-plugin",
    "targets": {
        "build": {
            "executor": "@nrwl/webpack:webpack",
            "options": {
                // other options 
                "transformers": [
                    // register plugins here
                    {
                        "name": "@nestjs/swagger/plugin",
                        "options": {
                            "introspectComments": true
                        }
                    }
                ]
            },
        }
   }
}
```
